Give bots ability to auto-replace harvesters

reaperrr opened the pull request:
...when their number drops below the number of refineries.

~Depends on #15982.~

Closes #15962.

reaperrr committed:
Make bots deploy MCV on first tick

And use a boolean instead of counting ticks.

reaperrr committed:
Increase default scan interval for idle bot harvesters

Every 2 seconds (at default speed) should be enough.

reaperrr committed:
Add plumbing for bots auto-replacing harvesters

If their number drops below refinery count.

reaperrr committed:
Enable harvester replacement in official mods

pchote commented on OpenRA.Mods.Common/Traits/BotModules/HarvesterBotModule.cs:
```diff
@@ -99,6 +111,30 @@ void IBotTick.BotTick(IBot bot)
+			if (!Info.HarvesterTypes.Any())
+				return false;
+
+			if (AIUtils.CountBuildingByCommonName(Info.HarvesterFactoryTypes, player) == 0)
```
If you check the `unitBuilder.RequestedProductionCount` here instead then we should be able to drop the factory types completely - it'll queue a single harv, then wait until that is completed before doing anything else.  This simplifies the yaml config, and also saves perf.

reaperrr replied:
Done and moved the whole check inline into BotTick.

reaperrr committed:
Remove unused BuildUnit overload from UnitBuilderBotModule

Unused and didn't check things like Buildable, so better just remove it.

reaperrr committed:
Add BotDebug message for external unit build requests

For easier bot debugging of things like MCV- and harvester replacement.

reaperrr commented:
Updated. moved the 'should build' decision making inline into BotTick, and added a BotDebug message for external unit build requests in the unit builder module.

Since the pt hasn't been tagged yet, maybe we can still sneak this in?

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/168534c9b3fe99dbda6c59e8fd836cbd9a25c9ec...141bb6f8f911e108b9125fa4de656ae42314b0ed)
